### PR TITLE
[CPDLP-3282] NPQ MVP Prep - Review and add missing validations for services and models

### PR DESCRIPTION
**Pull request by mooktakim**
### Context

Ticket: https://dfedigital.atlassian.net.mcas.ms/browse/CPDLP-3282

### Changes proposed in this pull request

* Moved validations from service classes to the model related to the specific columns.
* Updated factories to be valid

### CODE REVIEW

File: spec/models/declaration_spec.rb [hunk 1]
```diff
@@ -16,6 +16,39 @@
...
+      before { subject.declaration_date = subject.application.schedule.applies_from.prev_week }
+
+      it "has a meaningful error" do
+        expect(subject).to be_invalid
+        expect(subject).to have_error(:declaration_date, :declaration_before_schedule_start, "Enter a '#/declaration_date' that's on or after the schedule start.")
+      end
+    end
+
+    context "when declaration_date is at the schedule start" do
+      let(:schedule_applies_from_date) { declaration_date }
```
Comment (ethax-ross) [inline spec/models/declaration_spec.rb:48]: This isn't used anywhere; I think this should be something like:

```
before { subject.declaration_date = subject.application.schedule.applies_from }
```

I'm surprised rubocop didn't flag it 🤔

File: app/services/valid_test_data_generators/applications_populater.rb [hunk 2]
```diff
@@ -39,9 +39,7 @@ def initialize(lead_provider:, cohort:, number_of_participants:, logger: Rails.l
 
     def create_participants!
```
Comment (ethax-ross) [inline app/services/valid_test_data_generators/applications_populater.rb:40]: Why is the applications populater changing as part of this?

File: app/models/application.rb [hunk 3]
```diff
@@ -176,4 +181,17 @@ def funding_eligibility(with_funded_place:)
 
     eligible_for_funding && (funded_place.nil? || funded_place)
   end
+
+  def validate_funded_place?
+    accepted? && errors.blank? && cohort&.funding_cap?
+  end
+
+  def eligible_for_funded_place
```
Comment (ethax-ross) [inline app/models/application.rb:189]: Is it worth having these shared validations in a concern so we're not duplicating them?

**Review by ethax-ross — COMMENTED** [review]
I think there are more validation rules that could be moved to the models; I'm not sure how far we want to go with it though... some feel like they would make sense, for example:

- application -> validate_permitted_schedule_for_course (accept)
- participant outcome -> completion_date_not_in_the_future (create)

I'm less sure about the more service-specific ones though, such as:

- application -> cannot_change_from_rejected (accept)
- application -> cannot_change_from_accepted (reject)

File: app/services/valid_test_data_generators/applications_populater.rb (hunk 2 above)
Comment (mooktakim) [reply app/services/valid_test_data_generators/applications_populater.rb:40]: As we're creating seed data for different cohorts, validations related to dates get triggered for declarations. I had to move the `travel_to` so we create declarations that are after `schedule.applies_from`.

File: app/models/application.rb (hunk 3 above)
Comment (mooktakim) [reply app/models/application.rb:189]: There is slight differences to both methods, including how the errors are added. I'm not sure if its worth splitting it out.

**Comment by mooktakim** [comment]
> I think there are more validation rules that could be moved to the models; I'm not sure how far we want to go with it though... some feel like they would make sense, for example:
> 
> * application -> validate_permitted_schedule_for_course (accept)
> * participant outcome -> completion_date_not_in_the_future (create)
> 
> I'm less sure about the more service-specific ones though, such as:
> 
> * application -> cannot_change_from_rejected (accept)
> * application -> cannot_change_from_accepted (reject)

The `completion_date_not_in_the_future` for participant outcome already exists.

`validate_permitted_schedule_for_course` will look to include this

File: app/services/valid_test_data_generators/applications_populater.rb (hunk 2 above)
Comment (ethax-ross) [reply app/services/valid_test_data_generators/applications_populater.rb:40]: Makes sense 👍 I think we need to keep some of the randomness here though; so previously this was creating participants on random dates over the last 3 years whereas now the seeds would all have the same created at?

Comment (mooktakim) [reply app/services/valid_test_data_generators/applications_populater.rb:40]: Sure, I've tried to maintain some randomness with this line:
https://github.com/DFE-Digital/npq-registration/pull/1544/files#diff-38f0cbe9d749dd85fe13a256c7c104dfc5a070dbf1a7e0e205b6c9b80df3ff19R61

**Review by ethax-ross — APPROVED** [review]
Looks good 👍 I'm not sure if we need to have a think about if we should be adding some of the more service-specific validations into the core models or not, but that could be a follow-up PR.

**Review by cwrw — APPROVED** [review]
thanks @mooktakim, I think might be worth testing the data generator on review app for the PO review incase it changes anything with the randomness, but looks good otherwise

**Pull request closed without merging**